pb/forms_test.go

```diff
@@ -0,0 +1,357 @@
+		expectedForm: `{
+	"Links": [
+		{
+			"Tsize": 0
+		}
+	]
+}`,
+	})
+}
+
+// same as above but with a Hash
+func TestForm_Links_Hash_some_Tsize_zero(t *testing.T) {
+	var size uint64 = 0
+	l := &PBLink{Hash: cidBytes, Tsize: &size}
+	verifyRoundTrip(t, testCase{
+		node:          &PBNode{Links: []*PBLink{l}},
+		expectedBytes: "120d0a090155000500010203041800",
+		expectedForm: `{
+	"Links": [
+		{
+			"Hash": "015500050001020304",
+			"Tsize": 0
+		}
+	]
+}`,
+	})
+}
+
+// this case is excluded from the spec and should not be accepted at
+// higher-level APIs as a Hash property is mandatory
+func TestForm_Links_Tsize_some(t *testing.T) {
+	var size uint64 = 1010
+	l := &PBLink{Tsize: &size}
+	verifyRoundTrip(t, testCase{
+		node:          &PBNode{Links: []*PBLink{l}},
+		expectedBytes: "120318f207",
+		expectedForm: `{
+	"Links": [
+		{
+			"Tsize": 1010
+		}
+	]
+}`,
+	})
+}
+
+// same as above but with a Hash
+func TestForm_Links_Hash_some_Tsize_some(t *testing.T) {
+	var size uint64 = 1010
+	l := &PBLink{Hash: cidBytes, Tsize: &size}
+	verifyRoundTrip(t, testCase{
+		node:          &PBNode{Links: []*PBLink{l}},
+		expectedBytes: "120e0a0901550005000102030418f207",
+		expectedForm: `{
+	"Links": [
+		{
+			"Hash": "015500050001020304",
+			"Tsize": 1010
+		}
+	]
+}`,
+	})
+}
+
+func verifyRoundTrip(t *testing.T, tc testCase) {
+	actualBytes, actualForm, err := nodeRoundTripToString(tc.node)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if actualBytes != tc.expectedBytes {
+		fmt.Printf(
+			"Expected bytes: [%v]\nGot: [%v]\n",
+			tc.expectedBytes,
+			actualBytes)
+		t.Error("Did not match")
+	}
+
+	if actualForm != tc.expectedForm {
+		fmt.Printf(
+			"Expected form: [%v]\nGot: [%v]\n",
+			tc.expectedForm,
+			actualForm)
+		t.Error("Did not match")
+	}
+}
+
+func nodeRoundTripToString(n *PBNode) (string, string, error) {
+	bytes, err := n.Marshal()
+	if err != nil {
+		return "", "", err
+	}
+	fmt.Printf("[%v]\n", hex.EncodeToString(bytes))
+	rt := new(PBNode)
+	if err := rt.Unmarshal(bytes); err != nil {
+		return "", "", err
+	}
+	str, err := json.MarshalIndent(cleanPBNode(rt), "", "\t")
+	if err != nil {
+		return "", "", err
+	}
+	return hex.EncodeToString(bytes), string(str), nil
+}
+
+// convert a PBLink into a map for clean JSON marshalling
+func cleanPBLink(link *PBLink) *map[string]interface{} {
+	if link == nil {
+		return nil
+	}
+	// this would be a bad pb decode
+	if link.XXX_unrecognized != nil {
+		panic("Got unexpected XXX_unrecognized")
+	}
+	nl := make(map[string]interface{})
+	if link.Hash != nil {
+		nl["Hash"] = hex.EncodeToString(link.Hash)
+	}
+	if link.Name != nil {
+		nl["Name"] = link.Name
+	}
+	if link.Tsize != nil {
+		nl["Tsize"] = link.Tsize
+	}
+	return &nl
+}
+
+// convert a PBNode into a map for clean JSON marshalling
+func cleanPBNode(node *PBNode) *map[string]interface{} {
+	// this would be a bad pb decode
+	if node.XXX_unrecognized != nil {
+		panic("Got unexpected XXX_unrecognized")
```

Comment:
would be a bit nicer if you passed `t *testing.T` to these helper funcs, so this could be a `t.Fatal`. that way, if this really does happen, you fail one test, instead of panicking which stops all tests.

moot point if this should truly never happen, though.